SafeExamBrowser.SystemComponents/VirtualMachineDetector.cs

```diff
@@ -37,27 +39,136 @@ public VirtualMachineDetector(ILogger logger, ISystemInfo systemInfo)
+			bool isVirtualMachine = false;
+
+			biosInfo = biosInfo.ToLower();
+			manufacturer = manufacturer.ToLower();
+			model = model.ToLower();
 
 			isVirtualMachine |= biosInfo.Contains("hyper-v");
 			isVirtualMachine |= biosInfo.Contains("virtualbox");
 			isVirtualMachine |= biosInfo.Contains("vmware");
+			isVirtualMachine |= biosInfo.Contains("ovmf");
+			isVirtualMachine |= biosInfo.Contains("edk ii unknown");  // qemu
 			isVirtualMachine |= manufacturer.Contains("microsoft corporation") && !model.Contains("surface");
 			isVirtualMachine |= manufacturer.Contains("parallels software");
 			isVirtualMachine |= manufacturer.Contains("qemu");
 			isVirtualMachine |= manufacturer.Contains("vmware");
 			isVirtualMachine |= model.Contains("virtualbox");
+			isVirtualMachine |= model.Contains("Q35 +");  // qemu
+
+			return isVirtualMachine;
+		}
+
+		private bool IsVirtualRegistry()
+		{
+			bool isVirtualMachine = false;
+
+			// check historic hardware profiles
+			RegistryKey hardwareConfig = Microsoft.Win32.Registry.LocalMachine.OpenSubKey("SYSTEM\\HardwareConfig");
+
+			if (hardwareConfig != null)
+			{
+				foreach (string configId in hardwareConfig.GetSubKeyNames())
+				{
+					RegistryKey configKey = Microsoft.Win32.Registry.LocalMachine.OpenSubKey($"SYSTEM\\HardwareConfig\\{configId}");
+
+					if (configKey == null)
+					{
+						continue;
+					}
+
+					// reconstruct the systemInfo.biosInfo string
+					string biosInfo = (string) configKey.GetValue("BIOSVendor") + " " + (string) configKey.GetValue("BIOSVersion");
+					string manufacturer = (string) configKey.GetValue("SystemManufacturer");
+					string model = (string) configKey.GetValue("SystemProductName");
+
+					isVirtualMachine |= IsVirtualSystemInfo(biosInfo, manufacturer, model);
+
+					// hardware information of profile throughout installation etc. 
+					RegistryKey computerIds = Microsoft.Win32.Registry.LocalMachine.OpenSubKey($"SYSTEM\\HardwareConfig\\{configId}\\ComputerIds");
+
+					if (computerIds == null)
+					{
+						continue;
+					}
+
+					foreach (string computerId in computerIds.GetSubKeyNames())
+					{
+						// e.g. manufacturer&version&sku&...
+						string computer = (string)computerIds.GetValue(computerId);
+						isVirtualMachine |= IsVirtualSystemInfo(computer, computer, computer);
+					}
+				}
+			}
+
+			// check Windows timeline caches for current hardware config
+			RegistryKey deviceCache = Microsoft.Win32.Registry.CurrentUser.OpenSubKey($"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\TaskFlow\\DeviceCache");
+			
+			if (deviceCache != null)
+			{
+				foreach (string cacheId in deviceCache.GetSubKeyNames())
+				{
+					RegistryKey cacheKey = Microsoft.Win32.Registry.CurrentUser.OpenSubKey($"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\TaskFlow\\DeviceCache\\{cacheId}");
+
+					if (cacheKey == null)
+					{
+						continue;
+					}
+
+					string currHostname = System.Environment.GetEnvironmentVariable("COMPUTERNAME").ToLower();
```

Comment:
Could you please verify that the `COMPUTERNAME` environment variable is always present on a Windows system? Otherwise, this line may lead to an application crash in case the call to `Environment.GetEnvironmentVariable()` returns `null` for non-existent variables.

Reply:
Good catch. I have fixed the issue to be careful